### PR TITLE
Add option to disable spawning

### PR DESCRIPTION
**Pull request by nexdrew**
Add support for `{spawn: false}`, as suggested [in yargs PR 240](https://github.com/bcoe/yargs/pull/240#discussion_r38243143) and [in os-locale issue 8](https://github.com/sindresorhus/os-locale/issues/8#issuecomment-135965626). The main use-case is for yargs, which is currently using a custom port of this library to get around undesired spawning, which can cause problems for Atom.

I assumed an explicit `false` value must be used. If you'd rather let any falsey value disable spawning, I could change this:

```js
if (cache || getEnvLocale() || opts.spawn === false) {
```

To this:

```js
if (cache || getEnvLocale() || (Object.prototype.hasOwnProperty.call(opts, 'spawn') && !opts.spawn)) {
```

Testing is a bit hacky and required an upgrade to `ava@0.1.0` (for `test.serial`), but it should work. The reason I had to introduce serial tests is due to mocking dependencies and manipulating `process.env`. I left the existing tests untouched so they still run concurrently.

One odd thing I found during testing the async API is that if `childProcess.execFile` experiences an error (meaning first argument in callback is valid), then the callback given to os-locale is not fired. I'm not sure if this is intended, but we could possibly fix by changing instances of this:

```js
if (err) {
  fallback();
  return;
}
```

To this:

```js
if (err) {
  fallback();
  cb(err, cache);
  return;
}
```

As is now, the PR does not change this behavior.

### CODE REVIEW

**Comment by nexdrew** [comment]
Just realized I forgot to update the readme. I will add a commit soon.

**Comment by bcoe** [comment]
@nexdrew this is looking good, my other concern though is that `spawn-sync` has a post-install step on Windows that is prone to breaking, and was causing some troubles on AppVeyor:

https://github.com/ForbesLindesay/spawn-sync/blob/master/postinstall.js

I wonder if we could make `exec-file-sync` an optional dependency, so that a dependent like yargs could avoid performing the post-install step?

**Comment by sindresorhus** [comment]
> One odd thing I found during testing the async API is that if childProcess.execFile experiences an error (meaning first argument in callback is valid), then the callback given to os-locale is not fired.

It's intentional. See: https://github.com/sindresorhus/os-locale/commit/94c7ae54fdacdec7a66db5bfdb050a6f3d977fda

> I wonder if we could make exec-file-sync an optional dependency, so that a dependent like yargs could avoid performing the post-install step?

Optional dependencies have problems too, as users think the whole install failed if the optional install failed, because of bad wording by npm. I'd suggest we just remove the polyfill and only do spawning on Node 0.12+.

File: package.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
     "lcid": "^1.0.0"
   },
   "devDependencies": {
-    "ava": "0.0.4",
+    "ava": "0.1.0",
```
Comment (sindresorhus) [inline package.json]:
```
"ava": "*",
```

File: readme.md [hunk 2]
```diff
@@ -14,6 +14,11 @@ $ npm install --save os-locale
 
 ## Usage
 
+### Async
+
+Provide a callback that will be executed when the operating system's
+locale has been resolved.
```
Comment (sindresorhus) [inline readme.md]: Don't do hard-wrapping. Applies to the rest of the readme too.

File: test.js [hunk 3]
```diff
@@ -1,6 +1,26 @@
 'use strict';
 var test = require('ava');
 var osLocale = require('./');
+var envVars = ['LC_ALL', 'LANGUAGE', 'LANG', 'LC_MESSAGES'];
+var expectedFallback = 'en_US';
+
+function reloadOsLocale() {
+	delete require.cache[require.resolve('./')];
```
Comment (sindresorhus) [inline test.js]: You could use https://github.com/sindresorhus/clear-require or https://github.com/sindresorhus/require-uncached here.

File: test.js [hunk 4]
```diff
@@ -20,3 +40,72 @@ test('sync', function (t) {
...
+		restoreEnvVars(beforeTest);
+		require('child_process').execFile = beforeTest.childProcessExecFile;
+		reloadOsLocale();
+	};
+
+	// test async method
+	osLocale(function (err, locale) {
+		console.log('Locale identifier:', locale);
+		afterTest();
+		t.assert(!err, err);
```
Comment (sindresorhus) [inline test.js]: `t.error`.

https://github.com/sindresorhus/ava#context

File: test.js [hunk 5]
```diff
@@ -20,3 +40,72 @@ test('sync', function (t) {
...
+		t.assert(locale === expectedFallback, 'Locale did not match expected fallback');
```
Comment (sindresorhus) [inline test.js]: `t.is`

File: test.js [hunk 6]
```diff
@@ -20,3 +40,72 @@ test('sync', function (t) {
...
+		t.assert(locale !== 'spawn_NOTALLOWED', 'Attempted to spawn subprocess');
```
Comment (sindresorhus) [inline test.js]: `t.not`

**Comment by nexdrew** [comment]
> I'd suggest we just remove the polyfill and only do spawning on Node 0.12+.

See PR #11

**Comment by sindresorhus** [comment]
Superb! :)

**Pull request closed without merging**